### PR TITLE
test: Declare the inputs of the nested expect_* tests

### PR DESCRIPTION
**Pull request by bor-p-s**
### Description

The nested `expect_*` tests run a `bazel` of their own against the real source tree, invisible to the outer Bazel. Two problems follow: a cached pass can survive the change it was meant to catch, and `local` keeps every one of the 61 tests out of the shared cache, so every build re-runs all of them.

Each test now declares what decides its outcome: a fingerprint of the actions its fixture and its dependencies run (an aspect in `collect_actions.bzl`), the toolchains a `--extra_toolchains` flag registers (a transition), and the repo `.bazelrc`, `MODULE.bazel`(`.lock`), `.bazelversion`, and the compiler/runner jars. `local` becomes `no-sandbox` + `no-remote-exec`.

Analysis fails if the fingerprint carries no rules_scala action -- a fingerprint that can't notice a rules change must not serve a stale pass. 11 tests hit this (`scala_proto`/`semanticdb` compiling through their own aspect, two plain `proto_library`/`java_library` targets, and two `test/strict_dependency/scala_proto_deps` tests whose shared `--extra_toolchains` puts them under the same config as `:some_proto`'s own custom aspect, which Bazel reports as an `ActionConflictException`). Each says why with `no_fingerprint_reason` and is tagged `external`.

A fixture in another package used to be `external` unconditionally, since the macro can't verify a caller's `data` covers every source. Now that's only true when `target` is a pattern with no single label to fingerprint (`test/twitter_scrooge_jdk11`'s `.../...`); a concrete label gets the same real fingerprint a same-package fixture gets. It still can't safely add the fixture's own build output to `data` for full source coverage -- these fixtures are routinely `tags = ["manual"]`, built to fail or to only succeed under a toolchain the nested `bazel` supplies, and this macro can't read that tag from another package's BUILD file. Same tradeoff a near-empty fingerprint already accepts.

Two fixtures that fail during analysis moved to `analysistest` (#1922, merged, this PR's base).

`test/expect_build_failure/README.md` has the rest.

### Motivation

Removes a risk (a regression a nested test can no longer catch) and a cost (every nested test re-running on every build). The cross-package fixtures now cache: running them twice, the second run shows `(cached) PASSED`.

CI, same two steps against pre-branch master (#1918's build #6615) and this branch (#6637):

| | master | this branch |
|---|---|---|
| whole build | 1h19m05s | 1h16m48s |
| `bazel test //...` ubuntu | 2847s | 1484s |
| `bazel test //...` macOS | 3477s | 823s |
| `./test_rules_scala` (5 platform variants) | 3137-4712s | 3661-4600s |

Release impact: tests only.

### CODE REVIEW

File: test/expect_build_failure/collect_actions.bzl [hunk 1]
```diff
@@ -0,0 +1,100 @@
+"""Serializes the actions a target would run, so they can key a nested test.
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl:1]: Wow, I start to get it

File: test/expect_build_failure/collect_actions.bzl [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+# Every action this ruleset runs invokes a tool built from here, so a fingerprint
+# without it describes a target the rules never touch.
+_RULESET_TOOLS = "src/java/io/bazel/rulesscala"
+
+def _exposed_impl(ctx):
+    lines = sorted(ctx.attr.target[0][_ActionsInfo].lines.to_list())
+    if not [line for line in lines if _RULESET_TOOLS in line]:
+        fail(("%s runs no rules_scala action, so it cannot tell whether the rules " +
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl:86]: 🧠

File: test/strict_dependency/scala_proto_deps/BUILD [hunk 3]
```diff
@@ -85,12 +85,14 @@ expect_build_failure_test(
...
     expect = ["expected_stamp_by_convention.txt"],
+    no_fingerprint_reason = "under this toolchain combination collides with custom_phase_stamping_test below on :some_proto's aspect application (see that test's own no_fingerprint_reason)",
     target = ":uses_transitive_scala_proto",
 )
 
 expect_build_failure_test(
     name = "custom_phase_stamping_test",
     build_args = ["--extra_toolchains=//test/toolchains:ast_plus_one_deps_strict_deps_error,//test/strict_dependency/scala_proto_deps:stamp_by_convention_toolchain"],
     expect = ["expected_custom_phase_stamping.txt"],
+    no_fingerprint_reason = "fingerprinting this target under the same --extra_toolchains as stamp_by_convention_test above analyses :some_proto under two different aspect stacks (scala_proto_aspect alone vs. scala_proto_aspect + this package's custom_stamping_apsect) in the same configuration, which Bazel reports as an ActionConflictException on :some_proto's generated jar",
```
Comment (PawelLipski) [inline test/strict_dependency/scala_proto_deps/BUILD]: Typo: `apsect`

Comment (PawelLipski) [reply test/strict_dependency/scala_proto_deps/BUILD]: (oh, pre-existing, pls rename then)

File: test/strict_dependency/BUILD [hunk 4]
```diff
@@ -7,6 +7,7 @@ load("//test/expect_build_failure:expect_build_failure.bzl", "expect_build_failu
 expect_build_success_test(
     name = "plus_one_deps_only_works_for_java_info_targets_test",
     build_args = ["--extra_toolchains=//test_expect_failure/plus_one_deps:plus_one_deps"],
+    no_fingerprint_reason = "test_proto is a plain proto_library, so it runs no rules_scala action to fingerprint",
```
Comment (PawelLipski) [inline test/strict_dependency/BUILD:10]: Nit: would it make sense to add an attr to allow for no rules_scala actions in the given test? although AFAICS it's just these 2 tests, so maybe not worth it

File: test/twitter_scrooge_jdk11/BUILD [hunk 5]
```diff
@@ -16,12 +16,6 @@ expect_build_success_test(
         "--java_runtime_version=remotejdk_11",
         "--tool_java_runtime_version=remotejdk_11",
     ],
-    # The built target's sources live in another package, so the macro's
-    # automatic glob of this package doesn't cover them for staleness. Depend
-    # on a public target that transitively covers most of that package's
-    # scrooge sources (:twitter_scrooge_tests itself is package-private) so a
-    # source change still invalidates this test's cache key.
-    data = ["//test/src/main/scala/scalarules/test/twitter_scrooge:justscrooges_scala_and_java"],
```
Comment (PawelLipski) [inline test/twitter_scrooge_jdk11/BUILD:24]: This is now redundant right? as `fixture_actions` cover it anyway?

Or actually, `fixture_actions` don't "include" targets as such in any sense, right? only text dumps of actions from the transitive closure of the `target`?

File: test/expect_build_failure/nested_bazel.sh [hunk 6]
```diff
@@ -91,6 +91,16 @@ _nested_bazel_common_opts=()
 # then serialize on Bazel's output-base lock (each inner `bazel --batch` waits for
 # the lock rather than failing), which keeps the extracted external repos warm and
 # avoids multiplying ~1GB of Scala jars across a separate output base per test.
+_nested_bazel_home_hint() {
+  if [[ -n "${RULES_SCALA_NESTED_BAZEL_USE_REAL_HOME:-}" ]]; then
+    return
+  fi
+  echo "note: the nested \`bazel\` ran with a scrubbed HOME, so your ~/.bazelrc was" >&2
```
Comment (PawelLipski) [inline test/expect_build_failure/nested_bazel.sh:98]: Actually, is this fix specifically needed for this PR? or would it already be useful before (just the error not surfaced)?

Comment (PawelLipski) [reply test/expect_build_failure/nested_bazel.sh:98]: Coz scrubbed HOME was already there before (?)

File: test/expect_build_failure/expect_build_failure.bzl [hunk 7]
```diff
@@ -318,7 +471,9 @@ def expect_test_success_test(
         expect = [],
         reject = [],
         size = "large",
-        tags = ["local", "requires-network"],
+        tags = ["no-sandbox", "requires-network"],
```
Comment (PawelLipski) [inline test/expect_build_failure/expect_build_failure.bzl:479]: `local` is uncached but `no-sandbox` is, right?

File: test/classpath_resource_conflict/BUILD [hunk 8]
```diff
@@ -9,9 +9,8 @@ package(default_testonly = 1)
 # expect_build_success_test and assert the warning is surfaced. (See that macro's
 # docstring for how it keeps a cached action from hiding the warning.)
 #
-# `target` lives in another package, so its sources aren't tracked as this test's
-# runfiles (see expect_build_failure.bzl) and an edit there won't force a re-run;
-# the fixture is long-standing and rarely touched.
+# `target` lives in another package, so the macro tags this test `external`
+# (see expect_build_failure.bzl) and it re-runs on every invocation.
```
Comment (PawelLipski) [inline test/classpath_resource_conflict/BUILD]: Hmm okay - so this test is **silently** un-cached right (b/c the reason isn't lacking rules_scala in fingerprint but cross-package tested target)?

Maybe there should be an explicit opt-out boolean attr for these cases? like, mostly to increase visibility; not insisting tho

File: test/expect_build_failure/collect_actions.bzl [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+The fixture is analysed under the toolchains the nested `bazel` registers, so a
+change in one of those lands in the command lines too.
+
+Command lines carry output paths, and those name the configuration they were
+built in, so the fingerprint would otherwise differ under every flag the outer
+build happens to set and re-run the test for nothing. The configuration segment
+is dropped, in the spirit of Bazel's own path mapping. What is left out: actions
+that write a file rather than run a command line (`ctx.actions.write`, template
+expansion) contribute their mnemonic and nothing else, so a change to what such
+an action writes does not reach the fingerprint.
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl:18]: Hmmm and what if the **contents** of input files that the fingerprinted actions consume change?

Like, if `ScalaCompile` read a file (defined here in rules_scala, not in a project that consumes rules_scala) with the content `foobar`, and then it became `barfoo`? does it make sense?  I think the action dump won't change in such case. Although this is probably a niche corner case (as most file inputs are coming from the user's project, not from rules_scala)

File: test/expect_build_failure/collect_actions.bzl [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+def _without_configuration(line):
+    # Output paths carry the configuration (`bazel-out/k8-fastbuild-ST-1234/`),
+    # so the same fixture yields a different fingerprint under every flag the
+    # outer build happens to set, and every such build re-runs the test. The
+    # nested `bazel` does not inherit those flags, so drop the segment.
+    parts = line.split("bazel-out/")
+    out = parts[0]
+    for part in parts[1:]:
+        slash = part.find("/")
+        out += "bazel-out/" + ("<cfg>" + part[slash:] if slash != -1 else part)
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl:35]: Why two cases with/without slash?

File: test/expect_build_failure/collect_actions.bzl [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    return out
+
+def _collect_impl(target, ctx):
+    # One line per action, keeping the arguments in order. Collecting the tokens
+    # separately would lose which value belongs to which flag: a depset of tokens
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl]: This is unclear 🤔  is it just about keeping the command line as an (ordered) list rather than an (unordered) set?

File: test/expect_build_failure/collect_actions.bzl [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    # does not change when a flag's value is swapped for one that some other flag
+    # already carries, and `unused_dependency_checker_mode = "error"` -> `"off"`
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl]: Also, the example is unclear 🤔 (why this one specifically)

File: test/expect_build_failure/collect_actions.bzl [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    implementation = _collect_impl,
+    attr_aspects = ["*"],
+)
+
+def _toolchains_transition_impl(_settings, attr):
+    # Analyse the fixture under the toolchains the nested `bazel` will register,
+    # so that changing one of them changes the command lines collected above.
+    # Whatever the outer build registered is replaced, not extended: the nested
+    # `bazel` gets these and nothing else.
+    return {"//command_line_option:extra_toolchains": attr.extra_toolchains}
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl:71]: Wow, is it possible to swap ALL the available toolchains to a new list so easily? 😯

File: test/strict_dependency/scala_proto_deps/BUILD (hunk 3 above)
Comment (bor-p-s) [reply test/strict_dependency/scala_proto_deps/BUILD]: done, missed that.

File: test/strict_dependency/BUILD (hunk 4 above)
Comment (bor-p-s) [reply test/strict_dependency/BUILD:10]: Yeah, doesn't worth it.

File: test/expect_build_failure/collect_actions.bzl [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    implementation = _toolchains_transition_impl,
+    inputs = [],
+    outputs = ["//command_line_option:extra_toolchains"],
+)
+
+# Every action this ruleset runs invokes a tool built from here, so a fingerprint
+# without it describes a target the rules never touch.
+_RULESET_TOOLS = "src/java/io/bazel/rulesscala"
+
+def _exposed_impl(ctx):
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl]: Nit: why "exposed" impl?

Comment (PawelLipski) [reply test/expect_build_failure/collect_actions.bzl]: (and not just e.g. `_fixture_actions_impl`)

File: test/expect_build_failure/README.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
+# Asserting that a build or a test fails
```
Comment (PawelLipski) [inline test/expect_build_failure/README.md:1]: Nit: before reading in detail, is both the markdown and bzl doc needed? as in, do they overlap in the content? do both have sth that the other doesn't?

Comment (PawelLipski) [reply test/expect_build_failure/README.md:1]: Or maybe a better question: how is the division b/w what's here and what's in bzl defined?

File: test/expect_build_failure/expect_build_failure.bzl [hunk 16]
```diff
@@ -115,10 +169,76 @@ def _nested_bazel_test(
     #     directory the nested `bazel <command>` must run in. See
     #     _nested_bazel_find_workspace in nested_bazel.sh.
     #   - nested_bazel.sh: sourced at runtime by the helper script.
+    # Everything the nested `bazel` reads and this test's own key would
+    # otherwise miss: the repo `.bazelrc` (CI steps append to it, so two steps
```
Comment (PawelLipski) [inline test/expect_build_failure/expect_build_failure.bzl:178]: Do CI steps still append to .bazelrc? wasn't CI bazelrc aligned with local recently in your PR? 🤔

File: test/expect_build_failure/expect_build_failure.bzl [hunk 17]
```diff
@@ -21,15 +21,67 @@ A label inside a `build_args`/`bazel_args` flag value (e.g.
...
+_SCALAC_JAR = "//src/java/io/bazel/rulesscala/scalac:scalac_deploy.jar"
+
+# The compiler jar above does not pull in the test runners, and a nested
+# `bazel test` runs through them, so their content decides its outcome too.
+_RUNNER_JARS = [
+    "//src/java/io/bazel/rulesscala/test_discovery:test_discovery.jar",
+    "//src/java/io/bazel/rulesscala/specs2:specs2_test_discovery.jar",
+    "//src/java/io/bazel/rulesscala/scala_test:librunner.jar",
+    "//src/java/io/bazel/rulesscala/coverage/instrumenter:instrumenter.jar",
+]
```
Comment (PawelLipski) [inline test/expect_build_failure/expect_build_failure.bzl:87]: Actually, how was the list of these jars determined? (as in, how do know we that these ones and not more/less)

File: test/classpath_resource_conflict/BUILD (hunk 8 above)
Comment (bor-p-s) [reply test/classpath_resource_conflict/BUILD]: Right catch — the comment was stale. This target is a concrete label (not a pattern) in another package, so it actually goes through `fixture_actions` and gets a real fingerprint now, it's not tagged `external`. Fixed the comment to say that.

Comment (PawelLipski) [reply test/classpath_resource_conflict/BUILD]: Okay... so only **patterns**  don't go through fingerprinting right? could you re-iterate what's the problem with them?

File: test/twitter_scrooge_jdk11/BUILD (hunk 5 above)
Comment (bor-p-s) [reply test/twitter_scrooge_jdk11/BUILD:24]: `fixture_actions` doesn't declare `target` as `data`, it dumps the action command lines (mnemonic + argv) from the transitive closure, so a content-only source edit that doesn't change the command line wouldn't be caught. Moot here anyway: this fixture is a `...` pattern in another package, so it's tagged `external`, not fingerprinted.

File: test/expect_build_failure/nested_bazel.sh (hunk 6 above)
Comment (bor-p-s) [reply test/expect_build_failure/nested_bazel.sh:98]: Yes, needed for this PR. Before this change, the nested `bazel` always read your real `~/.bazelrc`, so its content could silently affect the test's outcome without being part of the cache key. That was already true before this PR, but it didn't matter yet, because nothing checked whether these tests were cached correctly. This PR is exactly what starts checking that, so the undeclared `~/.bazelrc` had to go.

File: test/expect_build_failure/expect_build_failure.bzl (hunk 7 above)
Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl:479]: Correct. `local` opts the test out of remote/shared caching entirely. `no-sandbox` only skips the sandbox and still lets the result be cached, which is what we want here.

File: test/expect_build_failure/collect_actions.bzl (hunk 9 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl:18]: Good catch, and yes, real gap. An action's command line only carries a file's path, not its content, so a content-only change to a file rules_scala itself bakes in wouldn't move the fingerprint. This is exactly why `_SCALAC_JAR`/`_RUNNER_JARS` in `expect_build_failure.bzl` are declared as explicit `data` instead of relying on the fingerprint. If another such file ever shows up, the fix is the same: add it to `data` by hand. Not generalizing this (e.g. hashing every input) since it's rare and the per-case fix is cheap.

File: test/expect_build_failure/collect_actions.bzl (hunk 10 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl:35]: Added a comment: the no-slash branch is for a token that ends right at the configuration directory, nothing follows it, so there's no anchor to replace up to; it's left unscrubbed in that rare case.

File: test/expect_build_failure/collect_actions.bzl (hunk 11 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl]: Reworded. It's not about order, it's about pairing: several rules_scala flags share the same off/warn/error values, so a flat depset of individual tokens can't tell which flag holds which value, and can miss a value moving from one flag to another.

File: test/expect_build_failure/collect_actions.bzl (hunk 12 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl]: Picked that pair because it's real: `unused_dependency_checker_mode`, `strict_deps_mode`, and `compiler_deps_mode` all draw from the same off/warn/error set, so a value can genuinely migrate between them. Reworded the comment to name that instead of just one flag.

File: test/expect_build_failure/collect_actions.bzl (hunk 13 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl:71]: Yep. `//command_line_option:extra_toolchains` written by a transition replaces the whole flag value for that configuration, it doesn't merge with whatever the outer build registered.

File: test/expect_build_failure/collect_actions.bzl (hunk 14 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl]: Fair, renamed to `_fixture_actions_impl` to match the rule it implements.

File: test/expect_build_failure/README.md (hunk 15 above)
Comment (bor-p-s) [reply test/expect_build_failure/README.md:1]: Some overlap, yes. README is the entry point: pick between the two mechanisms, the caching rationale, what to delete later. The `expect_build_failure.bzl` docstring is the source of truth for the actual cache-key list, kept next to the code so it can't drift from it. The overlap you're pointing at (`ctx.actions.write` actions) is one fact stated in both places; happy to trim it down to one if you'd rather.

File: test/expect_build_failure/expect_build_failure.bzl (hunk 16 above)
Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl:178]: Yes, still true: `.bazelci/presubmit.yml` has several steps doing `echo ... >>.bazelrc` (`lockfile_mode=update`, `test_env=PATH`, the JDK 21 flags on one step). Checked just now.

File: test/expect_build_failure/expect_build_failure.bzl (hunk 17 above)
Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl:87]: Each jar is a test rule's default dependency for the path a nested `bazel test` actually runs through: `scala_junit_test.bzl:74` (JUnit runner), `scala.bzl:47` (Specs2 discovery), `scala_test.bzl:78` (Scala test runner), `common_attributes.bzl:78` (coverage instrumenter). Found by searching the repo for every default label under `src/java/io/bazel/rulesscala` that a test rule wires in.

Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl:87]: Added a comment on the list: no automatic check keeps it complete, it's kept in sync by hand, same tradeoff as _SCALAC_JAR.

File: test/twitter_scrooge_jdk11/BUILD (hunk 5 above)
Comment (PawelLipski) [reply test/twitter_scrooge_jdk11/BUILD:24]: Yes, the point was more general about `fixture_actions`, not just about `test/twitter_scrooge_jdk11` - still, such minor omission feels acceptable

File: test/expect_build_failure/collect_actions.bzl (hunk 9 above)
Comment (PawelLipski) [reply test/expect_build_failure/collect_actions.bzl:18]: Oh! okay
So not really such a corner case after all

File: test/expect_build_failure/README.md (hunk 15 above)
Comment (PawelLipski) [reply test/expect_build_failure/README.md:1]: Okay, I'll read both in more detail now and comment if needed

File: test/expect_build_failure/expect_build_failure.bzl (hunk 16 above)
Comment (PawelLipski) [reply test/expect_build_failure/expect_build_failure.bzl:178]: Oh okay I've missed that

File: test/classpath_resource_conflict/BUILD (hunk 8 above)
Comment (PawelLipski) [reply test/classpath_resource_conflict/BUILD]: (@bor-p-s in case you've missed - mostly out of curiosity, not critical)

File: test/twitter_scrooge_jdk11/BUILD (hunk 5 above)
Comment (PawelLipski) [reply test/twitter_scrooge_jdk11/BUILD:24]: > content-only source edit that doesn't change the command line wouldn't be caught

Oh okay, addressed elsewhere, I see now

File: test/expect_build_failure/collect_actions.bzl [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        slash = part.find("/")
+        # No slash means the token ends right at the configuration directory (no
+        # path below it), so there's nothing to anchor the replacement on; leave
+        # it as-is rather than guess where the configuration name ends.
+        out += "bazel-out/" + ("<cfg>" + part[slash:] if slash != -1 else part)
+    return out
+
+def _collect_impl(target, ctx):
+    # One line per action, joining mnemonic and argv so each action keeps its own
+    # flag-to-value pairing. Collecting individual tokens into one depset would
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl:40]: Wait, why are actually "tokens" (as in, shell tokens?) stored in a **depset**? isn't a depset meant to keep, well, deps? is it some off-the-label use?

File: test/classpath_resource_conflict/BUILD [hunk 19]
```diff
@@ -9,9 +9,9 @@ package(default_testonly = 1)
 # expect_build_success_test and assert the warning is surfaced. (See that macro's
 # docstring for how it keeps a cached action from hiding the warning.)
 #
-# `target` lives in another package, so its sources aren't tracked as this test's
-# runfiles (see expect_build_failure.bzl) and an edit there won't force a re-run;
-# the fixture is long-standing and rarely touched.
+# `target` lives in another package, but it's one concrete label, so the macro
```
Comment (PawelLipski) [inline test/classpath_resource_conflict/BUILD]: Comment slightly redundant (up to you if it's worth keeping)

File: test/expect_build_failure/expect_build_failure.bzl [hunk 20]
```diff
@@ -21,15 +21,71 @@ A label inside a `build_args`/`bazel_args` flag value (e.g.
...
+Editing `.bazelversion` re-runs the tests. What that file cannot catch is a CI
+step which leaves it alone and points bazelisk at another binary. One step does
+that (`last_green`), and it keeps its own cache key regardless, because it
+appends three flags to `.bazelrc`, which is on the list above.
+
+Outside the list is whatever the network serves while external repositories are
+fetched. `MODULE.bazel.lock` pins the versions, so what is exposed is a flaky
+download rather than a silent change in what is being tested. Catching an input
+nobody thought of would take a run with `--nocache_test_results` on the branch
+we merge into, which means a step in `.bazelci/presubmit.yml`; this change does
```
Comment (PawelLipski) [inline test/expect_build_failure/expect_build_failure.bzl:57]: Wait, is there an existing step like that? or is such a step only proposed here?

File: test/expect_build_failure/expect_build_failure.bzl [hunk 21]
```diff
@@ -115,10 +174,76 @@ def _nested_bazel_test(
...
+    # `target` may be a pattern (a wildcard like `//pkg/...`), which has no
+    # single configured target to read actions from and no single label a
+    # cross-package caller could declare as `data`. A concrete label doesn't
+    # have that problem either way, so only a cross-package *pattern* is
+    # unfixable without the caller's help.
+    fingerprint_target = fingerprint_target or target
+    fingerprint_target_is_pattern = "..." in fingerprint_target or fingerprint_target.endswith((":all", ":*"))
+
+    if no_fingerprint_reason or (cross_package_fixture and fingerprint_target_is_pattern):
+        # The package glob above cannot cross a Bazel package boundary, and a
```
Comment (PawelLipski) [inline test/expect_build_failure/expect_build_failure.bzl]: > package glob above

Does it refer to `code_under_test`? maybe state explicitly

File: test/expect_build_failure/expect_build_failure.bzl [hunk 22]
```diff
@@ -115,10 +174,76 @@ def _nested_bazel_test(
...
+        # elsewhere in this file.
+        fixture_actions(
+            name = "%s_fixture_actions" % name,
+            target = fingerprint_target,
+            extra_toolchains = extra_toolchains,
+            testonly = True,
+        )
+        data.append(":%s_fixture_actions" % name)
+
+    # `no-sandbox` rather than `local`: both run the nested `bazel` outside the
```
Comment (PawelLipski) [inline test/expect_build_failure/expect_build_failure.bzl]: Somewhat misplaced comment, `no-sandbox` isn't declared here

File: test/expect_build_failure/expect_build_failure.bzl [hunk 23]
```diff
@@ -143,10 +268,10 @@ def _nested_bazel_test(
         srcs = [_HELPER],
         args = args,
         data = deduped_data,
-        # Each of these runs a nested `bazel` that takes the shared nested output
-        # base's lock. Run them one at a time (`exclusive`) so they do not queue
-        # behind each other and blow the test timeout.
-        tags = tags + ["exclusive"],
+        # `exclusive`: each runs a nested `bazel` that takes the shared output
```
Comment (PawelLipski) [inline test/expect_build_failure/expect_build_failure.bzl]: Same here, misplaced comment

File: test/strict_dependency/scala_proto_deps/BUILD [hunk 24]
```diff
@@ -85,12 +85,14 @@ expect_build_failure_test(
     name = "stamp_by_convention_test",
     build_args = ["--extra_toolchains=//test/toolchains:ast_plus_one_deps_strict_deps_error,//test/strict_dependency/scala_proto_deps:stamp_by_convention_toolchain"],
     expect = ["expected_stamp_by_convention.txt"],
+    no_fingerprint_reason = "under this toolchain combination collides with custom_phase_stamping_test below on :some_proto's aspect application (see that test's own no_fingerprint_reason)",
```
Comment (PawelLipski) [inline test/strict_dependency/scala_proto_deps/BUILD:88]: Actually, how did you detect that tests like this can't have a fingerprint? does analysis fail?

File: test/expect_build_failure/collect_actions.bzl [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+    out = parts[0]
+    for part in parts[1:]:
+        slash = part.find("/")
+        # No slash means the token ends right at the configuration directory (no
+        # path below it), so there's nothing to anchor the replacement on; leave
+        # it as-is rather than guess where the configuration name ends.
+        out += "bazel-out/" + ("<cfg>" + part[slash:] if slash != -1 else part)
+    return out
+
+def _collect_impl(target, ctx):
```
Comment (PawelLipski) [inline test/expect_build_failure/collect_actions.bzl]: Same as before, rename to `_collect_actions_aspect_impl` per convention

File: test/expect_build_failure/README.md [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+That nested build reads the real source tree, which the outer Bazel cannot see,
+so the test has to declare by hand everything that decides its outcome. The
+macro does that for you; the docstring in `expect_build_failure.bzl` lists what
+goes into the cache key and what stays outside it.
+
+Nothing re-runs these tests without the result cache on a schedule. That would be
+the way to catch an input nobody thought to declare.
+
+## What to delete later
```
Comment (PawelLipski) [inline test/expect_build_failure/README.md:28]: Pls create relevant issues once this PR lands

Comment (PawelLipski) [reply test/expect_build_failure/README.md:28]: Or in fact, just move this section to GitHub issues (and only leave links with short summaries)

File: test/expect_build_failure/README.md [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+  time rather than being served a pass it cannot vouch for. A few tests do
+  that today, for one of four reasons: `scala_proto` generates through an
+  aspect of its own, whose actions this aspect cannot read; the `semanticdb`
+  fixtures only build toolchain deps, so they compile nothing at all; a
+  fixture targets a plain `proto_library`/`java_library`, which runs no
+  rules_scala action either; or fingerprinting two fixtures under the same
+  `--extra_toolchains` would analyse a shared dependency under two different
+  aspect stacks, which Bazel reports as an `ActionConflictException`.
+
+What the fingerprint does not see: an action that writes a file instead of
```
Comment (PawelLipski) [inline test/expect_build_failure/README.md:80]: Also mention the change in rules_scala-declared file inputs

File: test/expect_build_failure/README.md [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+running a command line (`ctx.actions.write`, template expansion) contributes its
+mnemonic and nothing else, which is 446 of the 5135 lines it collects today. A
```
Comment (PawelLipski) [inline test/expect_build_failure/README.md]: Hmm it's unclear what the specific numbers refer to

Is it from the fingerprint of some specific target?

File: test/expect_build_failure/README.md [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+change to what such an action writes leaves the fingerprint identical. Declaring
+those files instead would put output paths -- and with them the configuration --
```
Comment (PawelLipski) [inline test/expect_build_failure/README.md]: Also unclear 🤔  what _those_ refer to?

File: test/classpath_resource_conflict/BUILD (hunk 8 above)
Comment (bor-p-s) [reply test/classpath_resource_conflict/BUILD]: Right, only patterns. The problem: `fixture_actions`'s aspect needs one concrete target to read actions from, a pattern like `//pkg/...` has no single configured target. And there's no single label the caller could declare as `data` either, so there's no fallback. A concrete label doesn't have either problem.

File: test/expect_build_failure/collect_actions.bzl (hunk 18 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl:40]: Not shell tokens. Each action's mnemonic and args are joined into one string per action (line 48). The depset holds these strings, not individual words.

Depset here merges each action's line with the lines already collected from its dependencies, across the whole graph, without flattening the list at every level.

File: test/classpath_resource_conflict/BUILD (hunk 19 above)
Comment (bor-p-s) [reply test/classpath_resource_conflict/BUILD]: Agreed, trimmed to a cross-reference to the macro's docstring instead of restating the mechanism.

File: test/expect_build_failure/expect_build_failure.bzl (hunk 20 above)
Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl:57]: Only proposed, doesn't exist. Checked .bazelci/presubmit.yml: no --nocache_test_results step there.

File: test/expect_build_failure/collect_actions.bzl (hunk 18 above)
Comment (PawelLipski) [reply test/expect_build_failure/collect_actions.bzl:40]: Okay! so off-label use IIUC, but smart/justified

File: test/expect_build_failure/expect_build_failure.bzl (hunk 21 above)
Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl]: Yes. Renamed the reference to `code_under_test` explicitly in the comment.

File: test/expect_build_failure/expect_build_failure.bzl (hunk 22 above)
Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl]: Right on both. Consolidated into one comment at the actual `execution_tags = tags + [...]` line -- that's where `exclusive`/`no-remote-exec` get added and where the `no-sandbox` vs `local` tradeoff belongs, since `no-sandbox` is a caller default, not decided here.

File: test/expect_build_failure/expect_build_failure.bzl (hunk 23 above)
Comment (bor-p-s) [reply test/expect_build_failure/expect_build_failure.bzl]: Fixed the same way -- both comments got folded into the one at `execution_tags = tags + [...]` (see the reply above).

File: test/strict_dependency/scala_proto_deps/BUILD (hunk 24 above)
Comment (bor-p-s) [reply test/strict_dependency/scala_proto_deps/BUILD:88]: Yes, analysis fails -- that's exactly what happened when I tried fingerprinting this one: Bazel's analysis phase itself failed with a real `ActionConflictException` on `:some_proto`'s generated jar, not something the macro's own `fail()` guard caught. The two tests analyse `:some_proto` under different aspect stacks in the same configuration, and Bazel won't allow two different action graphs to produce the same output path.

File: test/expect_build_failure/collect_actions.bzl (hunk 25 above)
Comment (bor-p-s) [reply test/expect_build_failure/collect_actions.bzl]: Done.

File: test/expect_build_failure/README.md (hunk 27 above)
Comment (bor-p-s) [reply test/expect_build_failure/README.md:80]: Added: the same blind spot applies to any file rules_scala reads by path rather than content -- `_SCALAC_JAR`/`_RUNNER_JARS` are declared as explicit `data` for exactly that reason.

File: test/expect_build_failure/README.md (hunk 28 above)
Comment (bor-p-s) [reply test/expect_build_failure/README.md]: Removed.

File: test/expect_build_failure/README.md (hunk 29 above)
Comment (bor-p-s) [reply test/expect_build_failure/README.md]: Fixed.

File: test/expect_build_failure/README.md (hunk 26 above)
Comment (bor-p-s) [reply test/expect_build_failure/README.md:28]: Agreed, will make issues after this lands and link them here.

**Review by PawelLipski — APPROVED** [review]
Good job! it  was very non-trivial 😅
